src/services/inlayHints.ts

```diff
@@ -0,0 +1,291 @@
+/* @internal */
+namespace ts.InlayHints {
+
+    const maxHintsLength = 30;
+
+    export function provideInlayHints(context: InlayHintsContext): InlayHint[] {
+        const { file, program, span, cancellationToken, preferences } = context;
+        const sourceFileText = file.text;
+
+        const checker = program.getTypeChecker();
+        const result: InlayHint[] = [];
+
+        visitor(file);
+        return result;
+
+        function visitor(node: Node): true | undefined | void {
+            if (!node || node.getFullWidth() === 0) {
+                return;
+            }
+
+            switch (node.kind) {
+
+                case SyntaxKind.ModuleDeclaration:
+                case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.InterfaceDeclaration:
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.ClassExpression:
+                case SyntaxKind.FunctionExpression:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.ArrowFunction:
+                    cancellationToken.throwIfCancellationRequested();
+
+                    if (!textSpanIntersectsWith(span, node.pos, node.getFullWidth())) {

```

Comment:
I think this code is in the wrong place - you are likely walking more of the tree than you need. Wouldn't it be better to stop the walk at the very top of this function?